### PR TITLE
[100% review] Document .taskcluster.yml version 1

### PR DESCRIPTION
**Pull request by djmitche**
Thoughts on this documentation?

As written, I feel like it's kind of dumping users in the deep end: write a YAML file that takes this obscure Github JSON object and finagles it into this obscure Taskcluster Task JSON object.  Good luck!

But, I don't like what happened with the env vars, where they seemed to proliferate and weren't especially well-defined, and meant different things for different events.

Any thoughts on an elegant way to make users' lives easier without just creating an endless pit of utility functions?

### CODE REVIEW

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [inline docs/taskcluster-yml-v1.md]: slugiid

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [inline docs/taskcluster-yml-v1.md]: Nice explanation

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [inline docs/taskcluster-yml-v1.md]: aritbrary

**Review by imbstack — DISMISSED** [review]
I agree this feels like the deep end a bit. Honestly as a first cut it is probably helpful to give them lots of copy-pasta examples and maybe we can work on creating something easier from there?

**Review by imbstack — COMMENTED** [review]
Oops, 10% review, so changes not approved! But the ideas so far are.

**Comment by djmitche** [comment]
That's kind of what I was thinking -- let's get our feet wet with v1, and see what commonalities we come up with and add those later.  Then we have a nice firm "you can do anything" base.

**Comment by owlishDeveloper** [comment]
(I had to fuss with the branch a bit, because I forgot to rebase it against master)

I think it's better to get rid of the description of the `version 0` altogether, so I moved some things to the `version 1` file. I also added some examples, as I currently understand what the `version 1` should look like. 

The one thing I'm the most unsure about is environment variables. Why would we need utility functions for them?

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Will this involve a JSON-e rendering?  If so, we need to define what the context looks like for that rendering.  If not, then we need to say that the rendering only applies to the `tasks` property, and not the `policy` property.

I'm leaning toward not rendering right now, for simplicity - @imbstack what do you think?

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Instead of using `extra.github`, I think we should require users to do this themselves using JSON-e expressions.  `{$if: 'tasks_for == 'github-push' && event.head.branch == 'development', then: ..}`

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: I think we want to get away from supplying these values.  We have never changed them, and probably never will, so we might as well just tell our users what provisionerId / workerType to use.

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: ++

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: There's no more top-level metadata :)

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Let's try to get rid of this section..

**Review by djmitche — COMMENTED** [review]
> I think it's better to get rid of the description of the version 0 altogether, so I moved some things to the version 1 file. 

I think it makes sense for v0 to be short, but we'll have a lot of people still using v0 files for a while, and they will want some way to figure out what those files mean.  

> The one thing I'm the most unsure about is environment variables. Why would we need utility functions for them?

It may be difficult or verbose to generate some of those values directly in JSON-e from the webhook payload.  For example, I think GITHUB_HEAD_TAG is generated by splitting the `refs/tags/` off the start of the head ref.

We may not end up needing these at all, or maybe only a few.  I think we should implement a v1 without them, or with only a few, and then try it out for some of the Taskcluster projects.  We will likely find some common annoyances -- and we can write utility functions for those.

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [inline docs/taskcluster-yml-v1.md]: speciifies

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [reply docs/taskcluster-yml-v1.md]: Rendering that would be confusing/scary! I agree, let's just render what is in `tasks`.

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [reply docs/taskcluster-yml-v1.md]: Totally agreed.

**Review by imbstack — COMMENTED** [review]
This is looking great.

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: I'll make sure that's clear in the doc. I'll also change the structure a bit and move the Policies section up, so that the rendering and tasks definition became neighbours.

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: Sounds cool!

File: docs/taskcluster-yml-v1.md
Comment (imbstack) [inline docs/taskcluster-yml-v1.md]: I think maybe it would be nicer if this was written as

```
tasks:
  - $if: tasks_for == "github-push"
    then:
      ...
```

**Review by imbstack — COMMENTED** [review]
This looks great. Should we leave some more of the `version 0` bits around though since they won't change?

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: Oh, ok. I'll try that syntax

Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: Yeah, that's much nicer! ❤️

File: docs/taskcluster-yml-v0.md
Comment (djmitche) [inline docs/taskcluster-yml-v0.md]: ++

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: As a sub-heading of Policies, this can be `##`

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Let's be really precise here: The `tasks` property of `.taskcluster.yml` is rendered..".  Then users will know that they can say something like:

```yaml
tasks:
  $let
    x: 10
  in:
    ...
```

that is, `tasks` need not be a literal array, as long as it JSON-e renders to an array.

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Let's use `*` instead of backticks here, as emphasis: "You can view it as a *template*. The following *context* variables are provided:"

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: `{$if, then}` (brace before $)

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: Maybe "Examples"?

**Review by djmitche — COMMENTED** [review]
This looks good!  Some minor formatting bits below.  You might also want to look at the outline of the document to see if the headings make sense.  One good way to do that is `cat docs/taskclutser-yml-v1.md | grep '^#'` to see just the headings and their levels.

Anyway, IMHO it's time to get implementin'!

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: Do you mean rename this section or add moar examples?

Comment (djmitche) [reply docs/taskcluster-yml-v1.md]: Sorry, I mean rename the section.  That's lots of examples already :)

Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md]: I like "Task definition" too... In the future we might want to add some task-related information in there. Besides, I have an "Example" section at the end. So I am proposing "Task definition and examples"

Comment (djmitche) [reply docs/taskcluster-yml-v1.md]: Sure, sounds good.

**Comment by owlishDeveloper** [comment]
I started implementing, and the first question I have is: I have to validate the yaml according to the schema in the tests, don't I? My understanding is we're gonna have a different schema 🤔

**Comment by imbstack** [comment]
@owlishDeveloper which schema in particular? We can always make a new schema!

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [inline docs/taskcluster-yml-v1.md]: @djmitche @imbstack I have a question on this: are we still using webhooks or do we use Events API?

**Comment by owlishDeveloper** [comment]
My commits are gone 🙀

**Comment by djmitche** [comment]
Did you push something to this branch?  I pushed 89160ff within the last few weeks, but may have force-pushed and deleted something in the process.  Do you still have those changes locally?

**Comment by owlishDeveloper** [comment]
Yes, I have them locally!

**Comment by djmitche** [comment]
@owlishDeveloper when you get a chance, please push them to this branch.  I apologize for not seeing them earlier.  then we'll get this merged!

**Comment by owlishDeveloper** [comment]
ok. Will tend to this at lunch!

**Comment by owlishDeveloper** [comment]
There are merge conflicts, so will be sorting out in the evening.

**Comment by owlishDeveloper** [comment]
Done! I think it needs one final review tho

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: A short section on tags might be useful here, since it's not clear in the github documentation.

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md:69]: We should probably note here that your template should be careful to produce nothing for unrecognized events (a caution we failed to follow in Firefox's .taskcluster.yml)

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: this should be `tasks_for == "github-push" && event.head.branch == "master"`

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [inline docs/taskcluster-yml-v1.md]: This task, and the one below, need to specify a taskId :)

**Review by djmitche — COMMENTED** [review]
I'll make these updates shortly (and not force-push over your changes this time!)

File: docs/taskcluster-yml-v1.md
Comment (owlishDeveloper) [inline docs/taskcluster-yml-v1.md:110]: what does `[:10]` mean here?

**Review by owlishDeveloper — APPROVED** [review]

File: docs/taskcluster-yml-v1.md
Comment (djmitche) [reply docs/taskcluster-yml-v1.md:110]: It's a slice, and it takes the prefix of the ref to compare.  The whole expression is equivalent to `event.ref.startswith('refs/tags/')`, but there's no `startswith` operator in JSON-e.

Comment (owlishDeveloper) [reply docs/taskcluster-yml-v1.md:110]: wow, that's cool!

**Review by imbstack — APPROVED** [review]
Sweet. We probably shouldn't land this till we land the little fixes in the try-v1 branch, but otherwise looks great!

**Pull request merged**
